### PR TITLE
Param store config

### PR DESCRIPTION
**Pull request by jkuipers**
Here's a contribution to address #207.
This code provides Spring Cloud centralized configuration support for the AWS Parameter Store by creating a PropertySourceLocator at bootstrap time that reads hierarchical configuration parameters.
Similar to Spring Cloud Config and the Consul Config support, shared configuration and overrides with profile-specific properties are supported.
Encrypted values are decrypted.

Currenty there's no support for publishing EnvironmentChangedEvents when updates are made to the parameters that make up combined PropertySources:
AFAICT the only way to that efficiently (i.e. without continuously polling all properties) would involve integrating with CloudWatch events, which I thought to be too complex for an initial implementation.

I first added my BootstrapConfiguration to the existing autoconfiguration module, but that module has a hard dependency on the context module and configures things I don' personally care for when I just want the Parameter Store integration.
That would also prevent me from running an application locally while still retrieving its config from the Parameter Store.
I therefore added autoconfiguration for this new module to its starter instead.

I've provided some documentation and a simple test: coming up with all this was already a significant amount of work, so I'd like some feedback first before putting in any more.
Please let me know what you think!

### CODE REVIEW

**Comment by aemruli** [comment]
first of all thanks @jkuipers will have a look this week and comment on this.

**Comment by SaikiranDaripelli** [comment]
@aemruli Any update on this?

File: docs/src/main/asciidoc/spring-cloud-aws.adoc [hunk 1]
```diff
@@ -573,6 +585,86 @@ public class ApplicationService {
 }
 ----
 
+=== Integrating your Spring Cloud application with the AWS Parameter Store
+
+Spring Cloud provides support for centralized configuration, which can be read and made available as a regular Spring
+`PropertySource` when the application is started. The Parameter Store Configuration allows you to use this mechanism
+with the https://docs.aws.amazon.com/systems-manager/latest/userguide/systems-manager-paramstore.html[AWS Parameter Store].
+
+Simply add a dependency on the `pring-cloud-starter-aws-parameter-store-config` starter module to activate the support.
```
Comment (devinsba) [inline docs/src/main/asciidoc/spring-cloud-aws.adoc]: nit: spelling

**Comment by jkuipers** [comment]
Would be nice to get an update on this: BTW, I'm running (another version of) this code in production right now, so it's proving itself as I'm typing this ;)

**Comment by masayuki038** [comment]
I would like this too, please.  What is the current status?

**Comment by joshualo** [comment]
+1

**Comment by Jarec** [comment]
+1

**Comment by gamov** [comment]
This is a great addition to spring-cloud-aws, IMO.

File: spring-cloud-aws-parameter-store-config/src/main/java/org/springframework/cloud/aws/paramstore/AwsParamStorePropertySource.java [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+    @Override
+    public Object getProperty(String name) {
+        return properties.get(name);
+    }
+
+    private void getParameters(GetParametersByPathRequest paramsRequest) {
+        GetParametersByPathResult paramsResult = source.getParametersByPath(paramsRequest);
+        for (Parameter parameter : paramsResult.getParameters()) {
+            String key = parameter.getName().replace(context, "").replace('/', '.');
+            properties.put(key, parameter.getValue());
```
Comment (devinsba) [inline spring-cloud-aws-parameter-store-config/src/main/java/org/springframework/cloud/aws/paramstore/AwsParamStorePropertySource.java]: Might it be good to handle KMS encrypted parameters here?

Comment (jkuipers) [reply spring-cloud-aws-parameter-store-config/src/main/java/org/springframework/cloud/aws/paramstore/AwsParamStorePropertySource.java]: They're already decrypted: `withWithDecryption(true)` takes care of that, assuming your service has the necessary authorities to do so.

**Comment by dsyer** [comment]
Merged in 24e2f81. I also re-deployed 2.0.0.RC2 to include this commit.

**Comment by jkuipers** [comment]
Thanks for merging this Dave!

**Comment by mariaoks** [comment]
Is there a way to store params of application for multi tenants?

**Pull request closed without merging**